tsdb/chunkenc/histo.go

```diff
@@ -0,0 +1,711 @@
+	b := make([]byte, 2, 128)
+	return &HistoChunk{b: bstream{stream: b, count: 0}}
+}
+
+// Encoding returns the encoding type.
+func (c *HistoChunk) Encoding() Encoding {
+	return EncHisto
+}
+
+// Bytes returns the underlying byte slice of the chunk.
+func (c *HistoChunk) Bytes() []byte {
+	return c.b.bytes()
+}
+
+// NumSamples returns the number of samples in the chunk.
+func (c *HistoChunk) NumSamples() int {
+	return int(binary.BigEndian.Uint16(c.Bytes()))
+}
+
+func (c *HistoChunk) Compact() {
+	if l := len(c.b.stream); cap(c.b.stream) > l+chunkCompactCapacityThreshold {
+		buf := make([]byte, l)
+		copy(buf, c.b.stream)
+		c.b.stream = buf
+	}
+}
+
+// Appender implements the Chunk interface.
+func (c *HistoChunk) Appender() (Appender, error) {
+	it := c.iterator(nil)
+
+	// To get an appender we must know the state it would have if we had
+	// appended all existing data from scratch.
+	// We iterate through the end and populate via the iterator's state.
+	for it.Next() {
+	}
+	if err := it.Err(); err != nil {
+		return nil, err
+	}
+
+	a := &histoAppender{
+		b: &c.b,
+
+		schema:   c.schema,
+		posSpans: c.posSpans,
+		negSpans: c.negSpans,
+
+		t:               it.t,
+		cnt:             it.cnt,
+		zcnt:            it.zcnt,
+		tDelta:          it.tDelta,
+		cntDelta:        it.cntDelta,
+		zcntDelta:       it.zcntDelta,
+		posbuckets:      it.posbuckets,
+		negbuckets:      it.negbuckets,
+		posbucketsDelta: it.posbucketsDelta,
+		negbucketsDelta: it.negbucketsDelta,
+
+		sum:      it.sum,
+		leading:  it.leading,
+		trailing: it.trailing,
+
+		buf64: make([]byte, binary.MaxVarintLen64),
+	}
+	if binary.BigEndian.Uint16(a.b.bytes()) == 0 {
+		a.leading = 0xff
+	}
+	return a, nil
+}
+
+// TODO fix this
+func (c *HistoChunk) iterator(it Iterator) *histoIterator {
+	// Should iterators guarantee to act on a copy of the data so it doesn't lock append?
+	// When using striped locks to guard access to chunks, probably yes.
+	// Could only copy data if the chunk is not completed yet.
+	//if histoIter, ok := it.(*histoIterator); ok {
+	//	histoIter.Reset(c.b.bytes())
+	//	return histoIter
+	//}
+	return &histoIterator{
+		// The first 2 bytes contain chunk headers.
+		// We skip that for actual samples.
+		br:       newBReader(c.b.bytes()[2:]),
+		numTotal: binary.BigEndian.Uint16(c.b.bytes()),
+		t:        math.MinInt64,
+	}
+}
+
+// Iterator implements the Chunk interface.
+//func (c *HistoChunk) Iterator(it Iterator) Iterator {
+//	return c.iterator(it)
+//}
+
+type histoAppender struct {
+	b *bstream
+
+	// Meta
+	schema             int32
+	posSpans, negSpans []histogram.Span
+
+	// for the fields that are tracked as dod's
+	t                           int64
+	cnt, zcnt                   uint64
+	tDelta, cntDelta, zcntDelta uint64
+
+	posbuckets, negbuckets           []int64
+	posbucketsDelta, negbucketsDelta []int64 // these could probably be uints instead. if the counts go up, the delta will always be >=0
```

Comment:
That's true if we always start a new chunk if the histogram is reset. But perhaps we don't want to do this at some point in the future.

Also, due to the limited lifespan of everything on earth, the int64 will never overflow, so we can save us the hassle of converting between int and uint.